Envoy master rebase

jrajahalme opened the pull request:
Now that all our Envoy infra enhancements have been upstreamed, we can
change our Envoy dependency to the official Envoy repo.
    
Pushed a new builder image using the updated dependencies, changed the
Dockerfile to depend on it.

The interface and implementation details were refined during the
upstreaming process, which requires corresponding changes in the
cilium listener filter.

Compile all Envoy API protos, except for the ones that are in draft
status and do not compile (currently metrics and trace services).

Envoy API protobuf files were also reorganized which requires changes
on how we build the golang stubs, and to make that easier we also
restructure the generated file layout in the subdirectories of
pkg/envoy as follows:

- pkg/envoy/envoy: .pb.go files generated from Envoy APIs
- pkg/envoy/cilium: .pb.go files generated from Cilium filter protobuf APIs

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet commented on envoy/BUILD:
```diff
@@ -25,28 +30,28 @@ envoy_cc_binary(
 )
 
 # TODO: Replace has_services=1 with a new api_go_grpc_library target after rebasing to use data-plane-api's master.
 api_proto_library(
```
They also now have new Bazel targets, which may be worth looking into? `api_go_proto_library` and `api_go_grpc_library`

jrajahalme replied:
I tried them out but would have needed a lot more time to figure out if we can use them. Decided to not go for Bazel build of protos top go this time around.

rlenglet commented on envoy/BUILD:
```diff
@@ -92,6 +97,7 @@ envoy_cc_library(
         "@envoy//source/common/common:assert_lib",
         "@envoy//source/common/common:logger_lib",
         "@envoy//source/common/network:address_lib",
+	":cilium_bpf_metadata_cc",
```
Use spaces to indent.

jrajahalme replied:
Fixed.

rlenglet commented on envoy/Makefile:
```diff
@@ -28,63 +28,22 @@ CLANG_FORMAT ?= clang-format
 BAZEL_BUILD_OPTS =
 
-all: clean-bins envoy $(GO_TARGETS)
+all: clean-bins envoy api
```
Add the `api` target to `.PHONY`.

Also add the 15 other existing phony targets that are missing:
```
$ sed -n 's/^\([^ $]*\):.*$/\1/p' Makefile | sort -u | grep -v '\.PHONY' | while read target ; do grep -q '^\.PHONY:.*'$target Makefile || echo $target ; done
all
bazel-archive
bazel-clean-archive
bazel-restore
check
clean
clean-bins
debug
debug-tests
envoy
envoy-debug
fix
install
release
tests
veryclean
```

jrajahalme replied:
Added `force-non-root` to api. All others already have some variant of force, and they all are .PHONY.

rlenglet commented on envoy/Makefile:
```diff
@@ -28,63 +28,22 @@ CLANG_FORMAT ?= clang-format
 
-release: envoy-release $(GO_TARGETS)
+api: Makefile.api
+	make -f Makefile.api all
```
`make` -> `$(MAKE)`

jrajahalme replied:
Fixed.

rlenglet commented on envoy/Makefile.api:
```diff
@@ -0,0 +1,66 @@
+# Copyright 2017 Authors of Cilium
```
2018

jrajahalme replied:
Fixed.

rlenglet commented on envoy/Makefile.api:
```diff
@@ -0,0 +1,66 @@
+	for path in $(PROTO_DIRS) ; do \
+		echo Compiling protos in $$path; \
+		$(PROTOC) -I $(ENVOY_API_PROTO_PATH) -I $(CILIUM_PROTO_PATH) $(PROTO_DEPS) --go_out=plugins=grpc$(GO_MAPPINGS):$(GO_OUT) $${path}*.proto; \
+	done
```
Add:
```
.PHONY: all
```

jrajahalme replied:
Added, thanks!

rlenglet commented on envoy/Makefile.api:
```diff
@@ -0,0 +1,66 @@
+GO_OUT = ../pkg/envoy
+
+# Skip draft Envoy APIs that do not compile.
+ENVOY_RAW_PROTOS := $(strip $(shell find -H $(ENVOY_API_PROTO_PATH)/envoy \( -name *.proto -and -not -path *service/trace/v2* -and -not -path *service/metrics/v2* \) -print))
```
If you listed explicitly all the *.proto files that we need in Cilium, you could then merge these Makefile rules back into`envoy/Makefile`:
```
ENVOY_RAW_PROTOS := blah.proto ...
$(ENVOY_RAW_PROTOS): envoy
```
I think that the only thing that's blocking merging this back is this line, because you decided to dynamically list those files.

jrajahalme replied:
Right, felt it was too much work to do that manually as the directory structure is wide and deep, and I would need to track the dependencies between the proto files manually.

jrajahalme committed:
envoy: Depend on upstream Envoy repo.

Now that all our Envoy infra enhancements have been upstreamed, we can
change our Envoy dependency to the official Envoy repo.

Pushed a new builder image using the updated dependencies, changed the
Dockerfile to depend on it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
ginkgo: Bump timeouts to 90 minutes

Ginkgo tests run out of time, when Envoy build added 20 minutes due to
caches not been updated yet.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Bump Bazel to 0.10.1

Upstream Envoy builds with this now.

Pushed a new builder image (cilium/cilium-builder:2018-02-16) using
the new bazel, changed the Dockerfile to depend on it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet reviewed (approved):
LGTM! I only have minor nitpicks.

jrajahalme committed:
envoy: Use V2 config for L7 filter as well.

This gets rid of the last bit of deprecated v1 config we were using.

Also restored the v1 config default for the bpf filter configuration
option "bpf_path".

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Review fixes.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme reviewed (commented):
Thanks for the review!

jrajahalme committed:
envoy: Rebase to upstream Envoy repo.

Now that we have upstreamed all new Envoy features that are needed to
run Envoy as a (more) transparent proxy with the Cilium bpf datapath,
we can rebase to the upstream Envoy repo.

The interface and implementation details were refined during the
upstreaming process, which requires corresponding changes in the
cilium listener filter.

Compile all Envoy API protos, except for the ones that are in draft
status and do not compile (currently metrics and trace services).

Envoy API protobuf files were also reorganized which requires changes
on how we build the golang stubs, and to make that easier we also
restructure the generated file layout in the subdirectories of
pkg/envoy as follows:

- pkg/envoy/envoy: .pb.go files generated from Envoy APIs
- pkg/envoy/cilium: .pb.go files generated from Cilium filter protobuf APIs

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
vendor: gogo-protobuf update.

More packages from github.com/gogo/protobuf are now needed to compile Envoy APIs.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
vendor: google.golang.org/genproto update.

Generated Envoy API proto files depend on these.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Docker builder update.

apt update for 16.10 did not work any more, testing with 17.10. Not
sure yet if runtime needs to be upgraded as well.

Use envoy/BAZEL_VERSION to automatially install the correct Bazel
version for the builder image. Use more separate steps in
Dockerfile.builder so that we don't need to start from scratch each
time.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Limit jobs to 1 on tests.

Saw out of memory error on test run on CI. Hope limiting number of
parallel jobs will help.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer reviewed (approved):
LGTM as janitor.
